api/src/main/java/io/minio/MinioClient.java

```diff
@@ -2839,27 +3056,331 @@ public void putObject(String bucketName, String objectName, InputStream stream,
+    SecureRandom ivSeed = new SecureRandom();
+    ivSeed.nextBytes(iv);
+
+    // Get CipherStream after encryption of input stream
+    CipherInputStream cipherInputStream = Crypto.encrypt(stream, dataEncryptionKey, dataEncryptionCipherMode,
+        iv);
+
+    // encrypt the plain data key using master key
+    byte[] dataKey = dataEncryptionKey.getEncoded();
+    byte[] encryptedDataKey = Crypto.encrypt(dataKey, masterKey, keyEncryptionCipherMode);
+
+    // Prepare data to be put to the object header
+    String encDataKey = BaseEncoding.base64().encode(encryptedDataKey);
+    String ivString = BaseEncoding.base64().encode(iv);
+
+    // Set the headermap with encryption related metadata
+    Map<String, String> headerMap = new HashMap<>();
+    headerMap.put("x-amz-meta-x-amz-key", encDataKey);
+    headerMap.put("x-amz-meta-x-amz-iv", ivString);
+    // matdesc is unused
+    headerMap.put("x-amz-meta-x-amz-matdesc", "{}");
+
+    // // Create byte array stream to hold encrypted stream
+    // ByteArrayOutputStream encryptedStream = new ByteArrayOutputStream();
+    //
+    // // Read bytes from cipher stream
+    // int numRead = 0;
+    // byte[] readBytes = new byte[8192];
+    // while ((numRead = cipherInputStream.read(readBytes)) != -1) {
+    // // write to byte array stream
+    // encryptedStream.write(readBytes, 0, numRead);
+    // }
+    //
+    // // Prepare input for put
+    // ByteArrayInputStream isToUpload = new ByteArrayInputStream(encryptedStream.toByteArray());
+
+    try {
+      // After padding the size changes, so isToUpload.available() instead of size passed from outside
+      putObject(bucketName, objectName, contentType, null, new BufferedInputStream(cipherInputStream),
+          headerMap);
+    } finally {
+      // Close used streams
+      cipherInputStream.close();
+      // encryptedStream.close();
+      // isToUpload.close();
+    }
+  }
 
   /**
    * Executes put object and returns ETag of the object.
    *
-   * @param bucketName   Bucket name.
-   * @param objectName   Object name in the bucket.
-   * @param length       Length of object data.
-   * @param data         Object data.
-   * @param uploadId     Upload ID of multipart put object.
-   * @param partNumber   Part number of multipart put object.
+   * @param bucketName
+   *          Bucket name.
+   * @param objectName
+   *          Object name in the bucket.
+   * @param length
+   *          Length of object data.
+   * @param data
+   *          Object data.
+   * @param uploadId
+   *          Upload ID of multipart put object.
+   * @param partNumber
+   *          Part number of multipart put object.
    */
   private String putObject(String bucketName, String objectName, int length,
-                           Object data, String uploadId, String contentType,
-                           int partNumber)
+      Object data, String uploadId, String contentType, int partNumber, Map<String, String> headerMap)
```

Comment:
remove `contentType` as we are receiving `headerMap` ie we would have `headerMap` updated with `contentType` prior to this call.